docs: rework introduction docs (#29062)
electron/electron: docs/tutorial/quick-start.md

# Quick Start

This guide will step you through the process of creating a barebones Hello World app in
Electron, similar to [`electron/electron-quick-start`][quick-start].

By the end of this tutorial, your app will open a browser window that displays a web page
with information about which Chromium, Node.js, and Electron versions are running.

[quick-start]: https://github.com/electron/electron-quick-start

## Prerequisites

To use Electron, you need to install [Node.js][node-download]. We recommend that you
use the latest `LTS` version available.

> Please install Node.js using pre-built installers for your platform.
> You may encounter incompatibility issues with different development tools otherwise.

To check that Node.js was installed correctly, type the following commands in your
terminal client:

```sh
node -v
npm -v
```

The commands should print the versions of Node.js and npm accordingly.

> **Note:** Since Electron embeds Node.js into its binary, the version of Node running

Reviewer: I think the guidance is to refer to the software as `Node.js` except when you're referring to the binary?
```suggestion
**Note:** Since Electron embeds Node.js into its binary, the version of Node.js running
```

your code is unrelated to the version running on your system.

[node-download]: https://nodejs.org/en/download/

## Create your application

### Scaffold the project

Electron apps follow the same general structure as other Node.js projects.
Start by creating a folder and initializing an npm package.

> ```sh npm2yarn

Reviewer: That looks like a nice magical directive...

Author: Yep, works out of the box with Docusaurus. Will be a no-op on the current website.

```sh npm2yarn
npm start
```

> Note: This script tells Electron to run on your project's root folder. At this stage,
> your app will immediately throw an error telling you that it cannot find an app to run.

[advanced-installation]: ./installation.md
[package-scripts]: https://docs.npmjs.com/cli/v7/using-npm/scripts

### Run the main process

The entry point of any Electron application is its `main` script. This script controls the
**main process**, which runs in a full Node.js environment and is responsible for
controlling your app's lifecycle, displaying native interfaces, performing privileged
operations, and managing renderer processes (more on that later).

During execution, Electron will look for this script in the [`main`][package-json-main]
field of the app's `package.json` config, which you should have configured during the
[app scaffolding](#scaffold-the-project) step.

To initialize the `main` script, create an empty file named `main.js` in the root folder
of your project.

> Note: If you run the `start` script again at this point, your app will no longer throw
> any errors! However, it won't do anything yet because we haven't added any code into

> > `main.js` yet.

Reviewer: I do this all the time, but you should pick the `yet` you prefer and remove the other one.

are available should open a new one.

To implement this feature, listen for the `app` module's [`activate`][activate]
event, and call your existing `createWindow()` method if no browser windows are open.

Because windows cannot be created before the `ready` event, you should only listen for
`activate` events after your app is initialized. Do this by attaching your event listener
from within your existing `whenReady()` callback.

[activate]: ../api/app.md#event-activate-macos

```js
app.whenReady().then(() => {
  createWindow()

  app.on('activate', function () {
    if (BrowserWindow.getAllWindows().length === 0) createWindow()
  })
})
```

> Note: At this point, your window controls should be fully functional!

### Access Node.js from the renderer with a preload script

Now, the last thing to do is print out the version numbers for Electron and its
dependencies onto your web page.

Accessing this information is trivial to do in the main process through Node's
global `process` object. However, you can't just edit the DOM from the main
process because it has no access to the renderer's `document` context.

> They're in entirely different processes!

Reviewer: Maybe link to the process model doc here?

This is where attaching a **preload** script to your renderer comes in handy.
A preload script runs before the renderer process is loaded, and has access to both
renderer globals (e.g. `window` and `document`) and a Node.js environment.

Create a new script named `preload.js` as such:

```js
window.addEventListener('DOMContentLoaded', () => {
  const replaceText = (selector, text) => {
    const element = document.getElementById(selector)
    if (element) element.innerText = text
  }

  for (const dependency of ['chrome', 'node', 'electron']) {
    replaceText(`${dependency}-version`, process.versions[dependency])
  }
})
```

> The above code accesses Node.js' `process.versions` object and runs a basic `replaceText`

Reviewer: Avoid Node.js as a possessive since possessive with an `s` rules are confusing:

```suggestion
The above code accesses the Node.js `process.versions` object and runs a basic `replaceText`
```

helper function to insert the version numbers into the HTML document.

To attach this script to your renderer process, pass in the path to your preload script
to the `webPreferences.preload` option in your existing `BrowserWindow` constructor.

```js

> // import Node's 'path' module at the top of your file

Reviewer: I'd generally avoid using `import` due to the keyword of the same name.

```suggestion
// include the Node.js 'path' module at the top of your file
```

const path = require('path')

// modify your existing createWindow() function
function createWindow () {
  const win = new BrowserWindow({
    width: 800,
    height: 600,
    preload: path.join(__dirname, 'preload.js')
  })

  win.loadFile('index.html')
}
// ...
```

There are two Node.js concepts that are used here:

* The [`__dirname`][dirname] string points to the path of the currently executing script
  (in this case, your project's root folder).
* The [`path.join`][path-join] API joins multiple path segments together, creating a

>   combined path string that works across all platforms.

Reviewer: I think it would be useful to mention here that using this style of file path generation is recommended due to [differences in paths when packaging](https://github.com/electron/electron-packager/blob/v15.2.0/docs/faq.md#why-isnt-the-relative-path-in-my-app-code-working), and then I can finally start moving this FAQ entry into the Electron docs itself.

[dirname]: https://nodejs.org/api/modules.html#modules_dirname
[path-join]: https://nodejs.org/api/path.html#path_path_join_paths

### Bonus: Add functionality to your web contents

At this point, you might be wondering how to add more functionality to your application.

For any interactions with your web contents, you want to add scripts to your
renderer process. Because the renderer runs in a normal web environment, you can add a
`<script>` tag right before your `index.html` file's closing `</body>` tag to include
any arbitrary scripts you want:

```html
<script src="./renderer.js"></script>
```

The code contained in `renderer.js` can then use the same JavaScript APIs and tooling
you use for typical front-end development, such as using [`webpack`][webpack] to bundle

> and minify your code or using [React][react] to manage your user interfaces.

Reviewer: I think the extra `using` is unnecessary?
```suggestion
and minify your code or [React][react] to manage your user interfaces.
```

    <meta http-equiv="X-Content-Security-Policy" content="default-src 'self'; script-src 'self'">
    <title>Hello World!</title>
  </head>
  <body>
    <h1>Hello World!</h1>
    We are using Node.js <span id="node-version"></span>,
    Chromium <span id="chrome-version"></span>,
    and Electron <span id="electron-version"></span>.

    <!-- You can also require other files to run in this process -->
    <script src="./renderer.js"></script>
  </body>
</html>
```

```fiddle docs/fiddles/quickstart
```

To summarize all the steps we've done:

* We bootstrapped a Node.js application and added Electron as a dependency.
* We created a `main.js` script that runs our main process, which controls our app
  and runs in a Node.js environment. In this script, we used Electron's `app` and
  `BrowserWindow` modules to create a browser window that displays web content
  in a separate process (the renderer).
* In order to access certain Node.js functionality in the renderer, we attached
  a preload script to our `BrowserWindow` constructor.

## Package and distribute your application

The fastest way to distribute your newly created app is using
[Electron Forge](https://www.electronforge.io).

> 1. Import Electron Forge to your app folder, and use the `import` script to set up

Reviewer: ```suggestion
1. Add Electron Forge as a development dependency of your app, and use its `import` command to set up
```
